Tests/TestUtils/TestNetworkService.swift

```diff
@@ -0,0 +1,190 @@
+            && lhs.url.path.lowercased() == rhs.url.path.lowercased()
+            && lhs.httpMethod.rawValue == rhs.httpMethod.rawValue
+    }
+    
+    func countDownExpected(networkRequest: NetworkRequest) {
+        for expectedNetworkRequest in expectedNetworkRequests {
+            if areNetworkRequestsEqual(lhs: expectedNetworkRequest.key, rhs: networkRequest) {
+                expectedNetworkRequest.value.countDown()
+            }
+        }
+    }
+
+    /// Starts the deadline timer for the given `NetworkRequest`, requiring all of its expected responses to have completed before the allotted time given in `timeout`.
+    ///
+    /// Note that it only sets the timer for the first `NetworkRequest` instance satisfying `areNetworkRequestsEqual`, using a dictionary backing.
+    /// This method is not recommended for instances where:
+    /// 1. Mutliple `NetworkRequest` instances would satisfy `areNetworkRequestsEqual` and all of them need the deadline timer started
+    /// 2. Order of the deadline timer application is important
+    private func awaitFor(networkRequest: NetworkRequest, timeout: TimeInterval) -> DispatchTimeoutResult? {
+        for expectedNetworkRequest in expectedNetworkRequests {
+            if areNetworkRequestsEqual(lhs: expectedNetworkRequest.key, rhs: networkRequest) {
+                return expectedNetworkRequest.value.await(timeout: timeout)
+            }
+        }
+
+        return nil
+    }
+
+    /// Returns all of the original outgoing `NetworkRequest`s satisfying `areNetworkRequestsEqual(lhs:rhs:)`.
+    func getSentNetworkRequestsMatching(networkRequest: NetworkRequest) -> [NetworkRequest] {
+        var matchingRequests: [NetworkRequest] = []
+        for request in sentNetworkRequests {
+            if areNetworkRequestsEqual(lhs: request.key, rhs: networkRequest) {
+                return request.value
+            }
+        }
+
+        return matchingRequests
+    }
+    
+    // MARK: Network request response helpers
+    /// Sets the `HttpConnection` response connection for a given `NetworkRequest`
+    func setResponseFor(networkRequest: NetworkRequest, responseConnection: HttpConnection?) {
+        networkResponses[networkRequest] = responseConnection
+    }
+    
+    // MARK: Assertion helpers
+    /// Set the expected number of times a NetworkRequest should be seen.
+    ///
+    /// - Parameters:
+    ///   - url: The URL for which to set the expectation
+    ///   - httpMethod: the `HttpMethod` for which to set the expectation, along with the `url`
+    ///   - count: how many times a request with this url and httpMethod is expected to be sent, by default it is set to 1
+    /// - See also:
+    ///     - assertNetworkRequestsCount()
+    ///     - getNetworkRequestsWith(url:httpMethod:)
+    func setExpectationForNetworkRequest(url: String, httpMethod: HttpMethod, expectedCount: Int32 = 1, file: StaticString = #file, line: UInt = #line) {
```

Comment:
any strong reason why some of these APIs do not use NetworkRequest param as the other APIs?